subprojects/gstreamer/plugins/tracers/gstframerate.c

```diff
@@ -0,0 +1,299 @@
+          "max", G_TYPE_INT, G_MAXINT,
+          NULL),
+      "ts", GST_TYPE_STRUCTURE, gst_structure_new_static_str ("value",
+          "type", G_TYPE_GTYPE, G_TYPE_UINT64,
+          "description", G_TYPE_STRING, "ts when the framerate has been logged",
+          "min", G_TYPE_UINT64, G_GUINT64_CONSTANT (0),
+          "max", G_TYPE_UINT64, G_MAXUINT64,
+          NULL),
+      "rendered", GST_TYPE_STRUCTURE, gst_structure_new_static_str ("value",
+          "type", G_TYPE_GTYPE, G_TYPE_UINT64,
+          "description", G_TYPE_STRING, "number of rendered frames",
+          "min", G_TYPE_UINT64, G_GUINT64_CONSTANT (0),
+          "max", G_TYPE_UINT64, G_MAXUINT64,
+          NULL),
+      "dropped", GST_TYPE_STRUCTURE, gst_structure_new_static_str ("value",
+          "type", G_TYPE_GTYPE, G_TYPE_UINT64,
+          "description", G_TYPE_STRING, "number of dropped frames",
+          "min", G_TYPE_UINT64, G_GUINT64_CONSTANT (0),
+          "max", G_TYPE_UINT64, G_MAXUINT64,
+          NULL),
+      "average", GST_TYPE_STRUCTURE, gst_structure_new_static_str ("value",
+          "type", G_TYPE_GTYPE, G_TYPE_FLOAT,
+          "description", G_TYPE_STRING, "Average frame rate",
+          NULL),
+      "current", GST_TYPE_STRUCTURE, gst_structure_new_static_str ("value",
+          "type", G_TYPE_GTYPE, G_TYPE_FLOAT,
+          "description", G_TYPE_STRING, "Number of frames rendered per time unit",
+          NULL),
+      "drop rate", GST_TYPE_STRUCTURE, gst_structure_new_static_str ("value",
+          "type", G_TYPE_GTYPE, G_TYPE_FLOAT,
+          "description", G_TYPE_STRING, "Number of frames dropped per time unit",
+          NULL),
+      NULL);
+  /* *INDENT-ON* */
+
+  GST_OBJECT_FLAG_SET (tr_framerate, GST_OBJECT_FLAG_MAY_BE_LEAKED);
+}
+
+static void
+do_element_post_message_pre (GstFramerateTracer *self, GstClockTime ts,
+    GstElement *element, GstMessage *message)
+{
+  GstElement *peer_element;
+
+  if (!is_target_video_sink (element))
+    return;
+
+  if (GST_MESSAGE_TYPE (message) == GST_MESSAGE_QOS) {
+    GstFormat format;
+    guint64 rendered, dropped;
+
+    gst_message_parse_qos_stats (message, &format, &rendered, &dropped);
+    if (format != GST_FORMAT_UNDEFINED) {
+      if (rendered != -1)
+        g_atomic_int_set (&self->frames_rendered, rendered);
+
+      if (dropped != -1)
+        g_atomic_int_set (&self->frames_dropped, dropped);
+    }
+  }
+}
+
+static void
+gst_framerate_tracer_init (GstFramerateTracer *self)
+{
+  GstTracer *tracer = GST_TRACER (self);
+
+  /* only trace pipeline framerate by default */
+  self->flags = GST_FRAMERATE_TRACER_FLAG_PIPELINE;
+  self->pid = getpid ();        // TODO: Not portable, i.e. Windows.
+
+  self->video_sink = NULL;
+
+  /* Init counters */
+  self->frames_rendered = 0;
+  self->frames_dropped = 0;
+  self->last_frames_rendered = G_GUINT64_CONSTANT (0);
+  self->last_frames_dropped = G_GUINT64_CONSTANT (0);
+  self->max_fps = -1;
+  self->min_fps = -1;
```

Comment:
sorry, I asked for a change that have broken

```c
  if (rr < self->min_fps) {
    self->min_fps = rr;
    GST_DEBUG_OBJECT (self, "Updated min-fps to %f", rr);
  }
```

can you init here with `G_MAXDOUBLE` ?

Reply:
or bring back
```c
  if (self->min_fps == -1 || rr < self->min_fps) {
    self->min_fps = rr;
    GST_DEBUG_OBJECT (self, "Updated min-fps to %f", rr);
  }
```
(but only for that block)

Reply:
brought it back, I prefer to have it as similar to fpsdisplaysink